src/signer.ts

```diff
@@ -0,0 +1,416 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as crypto from 'crypto';
+import * as dateFormat from 'date-and-time';
+import * as http from 'http';
+import * as url from 'url';
+import {encodeURI, qsStringify, objectEntries} from './util';
+
+interface GetCredentialsResponse {
+  client_email?: string;
+}
+
+export interface AuthClient {
+  sign(blobToSign: string): Promise<string>;
+  getCredentials(): Promise<GetCredentialsResponse>;
+}
+
+export interface BucketI {
+  name: string;
+}
+
+export interface FileI {
+  name: string;
+}
+
+export interface Query {
+  [key: string]: string;
+}
+
+interface GetSignedUrlConfigInternal {
+  expiration: number;
+  method: string;
+  extensionHeaders?: http.OutgoingHttpHeaders;
+  queryParams?: Query;
+  cname?: string;
+  contentMd5?: string;
+  contentType?: string;
+  bucket: string;
+  file?: string;
+}
+
+interface SignedUrlQuery {
+  generation?: number;
+  'response-content-type'?: string;
+  'response-content-disposition'?: string;
+}
+
+interface V2SignedUrlQuery extends SignedUrlQuery {
+  GoogleAccessId: string;
+  Expires: number;
+  Signature: string;
+}
+
+interface V4SignedUrlQuery extends V4UrlQuery {
+  'X-Goog-Signature': string;
+}
+
+interface V4UrlQuery extends SignedUrlQuery {
+  'X-Goog-Algorithm': string;
+  'X-Goog-Credential': string;
+  'X-Goog-Date': string;
+  'X-Goog-Expires': number;
+  'X-Goog-SignedHeaders': string;
+}
+
+export interface SignerGetSignedUrlConfig {
+  method: 'GET' | 'PUT' | 'DELETE' | 'POST';
+  expires: string | number | Date;
+  virtualHostedStyle?: boolean;
+  version?: 'v2' | 'v4';
+  cname?: string;
+  extensionHeaders?: http.OutgoingHttpHeaders;
+  queryParams?: Query;
+  contentMd5?: string;
+  contentType?: string;
+}
+
+export type GetSignedUrlResponse = string;
+
+export interface GetSignedUrlCallback {
+  (err: Error | null, url?: string): void;
+}
+type ValueOf<T> = T[keyof T];
+type HeaderValue = ValueOf<http.OutgoingHttpHeaders>;
+
+/*
+ * Default signing version for getSignedUrl is 'v2'.
+ */
+const DEFAULT_SIGNING_VERSION = 'v2';
+
+const SEVEN_DAYS = 604800;
+
+/**
+ * @const {string}
+ * @private
+ */
+export const PATH_STYLED_HOST = 'https://storage.googleapis.com';
+
+export class UrlSigner {
```

Comment:
Should we:

```diff
- UrlSigner
+ URLSigner
```

? Similar examples, `encodeURIComponent`, `compute.getVMs()`.

Reply:
refactored!

Reply:
Oh no, I forgot we have "getSignedUrl()" and not "getSignedURL()". I guess we'll just have a little of both :)